### PR TITLE
Add `assertThrowsExactly` assertion method

### PR DESCRIPTION
**Pull request by marcphilipp**
## Overview

Continuation of #2543.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc [hunk 1]
```diff
@@ -90,6 +90,9 @@ on GitHub.
 * `DynamicTests.stream()` can now consume `Named` input and will use each name-value pair
   as the display name and value for each generated dynamic test (see
   <<../user-guide/index.adoc#writing-tests-dynamic-tests-examples,User Guide>> for details).
+* New `assertThrowsExactly` method is introduced as a more strict version of `assertThrows`
+  allowing you to assert that the thrown exception has the exact specified class.
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:94]:
```suggestion
* New `assertThrowsExactly` method which is a more strict version of `assertThrows`
  that allows you to assert that the thrown exception has the exact specified class.
```

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc [hunk 2]
```diff
@@ -90,6 +90,9 @@ on GitHub.
 * `DynamicTests.stream()` can now consume `Named` input and will use each name-value pair
   as the display name and value for each generated dynamic test (see
   <<../user-guide/index.adoc#writing-tests-dynamic-tests-examples,User Guide>> for details).
+* New `assertThrowsExactly` method is introduced as a more strict version of `assertThrows`
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc:93]: This also needs to be moved to `release-notes-5.8.0-M2.adoc`.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrowsExactly.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.function.Supplier;
+
+import org.junit.jupiter.api.function.Executable;
+import org.junit.platform.commons.util.UnrecoverableExceptions;
+import org.opentest4j.AssertionFailedError;
+
+/**
+ * {@code AssertThrowsExactly} is a collection of utility methods that support asserting
+ * an exception of an expected type is thrown.
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrowsExactly.java:26]:
```suggestion
 * an exception of an exact type is thrown.
```

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/api/AssertThrowsExactlyAssertionsTests.java [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+				throw new NumberFormatException() {
+				};
+			});
+			expectAssertionFailedError();
+		}
+		catch (AssertionFailedError ex) {
+			assertMessageStartsWith(ex, "Unexpected exception type thrown ==> ");
+			assertMessageContains(ex, "expected: <java.lang.IllegalStateException>");
+			// As of the time of this writing, the class name of the above anonymous inner
+			// class is org.junit.jupiter.api.AssertionsAssertThrowsTests$2; however, hard
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/api/AssertThrowsExactlyAssertionsTests.java:250]:
```suggestion
			// class is org.junit.jupiter.api.AssertThrowsExactlyAssertionsTests$2; however, hard
```

I know it's nitpicking, but... technically it's true. 😇

**Review by sbrannen — CHANGES_REQUESTED** [review]
Looks good.

I only requested a few minor changes.

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc (hunk 2 above)
Comment (marcphilipp) [reply documentation/src/docs/asciidoc/release-notes/release-notes-5.8.0-M1.adoc]: Done.

**Review by sbrannen — APPROVED** [review]

**Pull request merged**